internal/validation/validation.go

```diff
@@ -0,0 +1,290 @@
+	}
+
+	// now we assume our tuple is well-formed, it's time to check
+	// the tuple against other model and type-restriction constraints
+
+	err := validateTuplesetRestrictions(typesys, tk)
+	if err != nil {
+		return err
+	}
+
+	objectType := tuple.GetType(tk.GetObject())
+	relation := tk.GetRelation()
+
+	hasTypeInfo, err := typesys.HasTypeInfo(objectType, relation)
+	if err != nil {
+		return err
+	}
+
+	if hasTypeInfo {
+		err := validateTypeRestrictions(typesys, tk)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// validateTuplesetRestrictions validates the provided TupleKey against tupleset restrictions.
+//
+// Given a rewrite definition such as 'viewer from parent', the 'parent' relation is known as the
+// tupleset. This method ensures the following are *not* possible:
+//
+// 1. `document:1#parent@folder:1#parent` (cannot evaluate/assign a userset value to a tupleset relation)
+// 2. `document:1#parent@*` (cannot evaluate/assign untyped wildcard to a tupleset relation (1.0 models))
+// 3. `document:1#parent@folder:*` (cannot evaluate/assign typed wildcard to a tupleset relation (1.1. models))
+func validateTuplesetRestrictions(typesys *typesystem.TypeSystem, tk *openfgapb.TupleKey) error {
+
+	objectType := tuple.GetType(tk.GetObject())
+	relation := tk.GetRelation()
+
+	isTupleset, err := typesys.IsTuplesetRelation(objectType, relation)
+	if err != nil {
+		return err
+	}
+
+	if !isTupleset {
+		return nil
+	}
+
+	rel, err := typesys.GetRelation(objectType, relation)
+	if err != nil {
+		return err
+	}
+
+	rewrite := rel.GetRewrite().GetUserset()
+
+	// tupleset relation involving a rewrite
+	if rewrite != nil && reflect.TypeOf(rewrite) != reflect.TypeOf(&openfgapb.Userset_This{}) {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("unexpected rewrite encountered with tupelset relation '%s#%s'", objectType, relation),
+			TupleKey: tk,
+		}
+	}
+
+	user := tk.GetUser()
+
+	// if a tupleset relation is related to an object but not a typed wildcard (e.g. 'user:*')
+	// then it is valid
+	if tuple.IsValidObject(user) && !tuple.IsTypedWildcard(user) {
+		return nil
+	}
+
+	// tupleset relation involving a wildcard (covers the '*' and 'type:*' cases)
+	if tuple.IsWildcard(user) {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("unexpected wildcard relationship with tupleset relation '%s#%s'", objectType, relation),
+			TupleKey: tk,
+		}
+	}
+
+	// tupleset relation involving a userset (e.g. object#relation) or a user_id (e.g. not a valid object)
+	if !tuple.IsValidObject(user) {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("unexpected user '%s' with tupleset relation '%s#%s'", user, objectType, relation),
+			TupleKey: tk,
+		}
+	}
+
+	return nil
+}
+
+// validateTypeRestrictions makes sure the type restrictions are enforced.
+// 1. If the tuple is of the form (user=person:bob, relation=reader, object=doc:budget), then the type "doc", relation "reader" must allow type "person".
```

Comment:
Can we write these tuples using the Zanzibar notation, i.e., `doc:budget#reader@person:bob`?

Reply:
Yep! Changed that.